### PR TITLE
Pin non-canonical snaps by revision but arch aware

### PR DESCRIPTION
**Pull request by addyess**
### Overview

Due to the need to continue installing snaps via revision, but in an arch specific way, the actions operator must install the appropriate snap revision according to the architecture of the running machine. 

### Details
First find the architecture of the current machine using dpkg, then install the appropriate snaps based on the arch and revision.

### CODE REVIEW

**Comment by addyess** [comment]
@ghislainbourgeois i believe this captures both architectures for those snaps.  I'm a little concerned about `juju-bundle` -- TBH -- hopefully it's a python snap or something?

File: src/bootstrap/index.ts [hunk 1]
```diff
@@ -245,10 +251,13 @@ async function run() {
...
         await snap(`install charmcraft --classic --channel=${charmcraft_channel}`);
 
-
-        await snap(`install jq ${fixed_revision_args("jq", "")}`);
-        await snap(`install juju-bundle --classic ${fixed_revision_args("juju-bundle", juju_bundle_channel)}`);
-        await snap(`install juju-crashdump --classic ${fixed_revision_args("juju-crashdump", juju_crashdump_channel)}`)
+        let arch = "";
+        const dpkg = _retryable_exec("dpkg");
+        const dpkg_output = {listeners:{stdout: (data: Buffer) => {arch += data.toString();}}};
+        await dpkg("--print-architecture", [], dpkg_output)
```
Comment (addyess) [inline src/bootstrap/index.ts:257]: reads the architecture of the machine into `arch`

File: src/bootstrap/index.ts [hunk 2]
```diff
@@ -178,10 +178,16 @@ async function snap_install(name: string, channel: string="", classic: boolean=t
...
+function fixed_revision_args(app:string, channel:string, arch:string): string{
     if (!channel) {
-        return `--revision=${pinning[app]}`
+        // Securely pin snap versions by arch
+        // snap versions are determined via API request to snapstore
+        // https://api.snapcraft.io/v2/snaps/info/<SNAP>?architecture=<ARCH>&fields=revision
+        const pinning = {
+            amd64: {"juju-bundle": 25, jq: 6, "juju-crashdump": 271},
+            arm64: {"juju-bundle": 25, jq: 8, "juju-crashdump": 272}
+        };
```
Comment (addyess) [inline src/bootstrap/index.ts:189]: I looked up the stable branches of these snaps, and i'm pinning to those revisions pre request from security.  These packages are not canonical owned snaps

**Review by addyess — COMMENTED** [review]
Annotations

**Review by mateoflorido — APPROVED** [review]
Nice! LGTM

**Pull request merged**